File: alljoyn/services/notification/cpp/src/NotificationService.cc
/******************************************************************************
 * Copyright (c) 2013-2014, AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 ******************************************************************************/

#include <algorithm>
#include <alljoyn/notification/NotificationSender.h>
#include <alljoyn/notification/NotificationService.h>
#include "NotificationConstants.h"
#include "Transport.h"
#include "NotificationProducerReceiver.h"
#include <alljoyn/notification/LogModule.h>

using namespace ajn;
using namespace services;
using namespace qcc;
using namespace nsConsts;

NotificationService* NotificationService::s_Instance(NULL);
uint16_t const NotificationService::NOTIFICATION_SERVICE_VERSION = 2;

NotificationService* NotificationService::getInstance()
{
    if (!s_Instance) {
        s_Instance = new NotificationService();
    }
    return s_Instance;
}

NotificationService::NotificationService()
{
}

NotificationService::~NotificationService()
{

}

uint16_t NotificationService::getVersion()
{
    return NOTIFICATION_SERVICE_VERSION;
}

NotificationSender* NotificationService::initSend(BusAttachment* bus, PropertyStore* store)
{
    QCC_DbgTrace(("NotificationService::initSend"));

    if (!bus) {
        QCC_DbgHLPrintf(("BusAttachment cannot be NULL"));
        return NULL;
    }

    if (!store) {
        QCC_DbgHLPrintf(("PropertyStore cannot be NULL"));
        return NULL;
    }

    Transport* transport = Transport::getInstance();
    if (transport->startSenderTransport(bus) != ER_OK) {
        return NULL;
    }

    MsgArg configArgs[1];
    MsgArg* configEntries;
    size_t configNum;
    QStatus status;

    if ((status = store->ReadAll(0, PropertyStore::READ, configArgs[0]))) {
        QCC_LogError(status, ("Error reading all in configuration data"));
    }

    if ((status = configArgs[0].Get(AJPARAM_ARR_DICT_STR_VAR.c_str(), &configNum, &configEntries))) {
        QCC_LogError(status, ("Error reading in configuration data"));
    }

    MsgArg appIdArg;
    for (size_t i = 0; i < configNum; i++) {
        char* keyChar;
        String key;
        MsgArg* variant;

        status = configEntries[i].Get(AJPARAM_DICT_STR_VAR.c_str(), &keyChar, &variant);
        if (status != ER_OK) {
            QCC_LogError(status, ("Error reading in configuration data"));
        }

        key = keyChar;

        if (key.compare("AppId") == 0) {
            appIdArg = *variant;
        }
    }

    if (status != ER_OK) {
        QCC_LogError(status, ("Something went wrong unmarshalling the propertystore."));
        return NULL;
    }

    if (appIdArg.typeId != ALLJOYN_BYTE_ARRAY) {
        QCC_DbgHLPrintf(("ApplicationId argument is not correct type."));
        return NULL;
    }

    transport->getNotificationProducerReceiver()->SetAppIdArg(appIdArg);

    return new NotificationSender(store);
}

QStatus NotificationService::initReceive(ajn::BusAttachment* bus, NotificationReceiver* notificationReceiver)
{
    if (!bus) {
        QCC_LogError(ER_BAD_ARG_1, ("BusAttachment cannot be NULL."));
        return ER_BAD_ARG_1;
    }

    if (!notificationReceiver) {
        QCC_LogError(ER_BAD_ARG_2, ("Could not set NotificationReceiver to null pointer"));
        return ER_BAD_ARG_2;
    }

    QCC_DbgPrintf(("Init receive"));

    Transport* transport = Transport::getInstance();
    transport->setNotificationReceiver(notificationReceiver);

    QStatus status;
    if ((status = transport->startReceiverTransport(bus)) != ER_OK) {
        transport->setNotificationReceiver(0);
    }

    return status;
}

void NotificationService::shutdownSender()
{
    QCC_DbgTrace(("Stop Sender"));
    Transport* transport = Transport::getInstance();
    transport->cleanupSenderTransport();
}

void NotificationService::shutdownReceiver()
{
    QCC_DbgTrace(("Stop Receiver"));
    Transport* transport = Transport::getInstance();
    transport->cleanupReceiverTransport();
}

void NotificationService::shutdown()
{
    QCC_DbgTrace(("Shutdown"));
    Transport* transport = Transport::getInstance();
    transport->cleanup();

    s_Instance = 0;
    delete this;
}

QStatus NotificationService::disableSuperAgent()
{
    QCC_DbgTrace(("Disabling SuperAgent"));
    Transport* transport = Transport::getInstance();
    return transport->disableSuperAgent();
}


BusAttachment* NotificationService::getBusAttachment()
{
    QCC_DbgTrace(("In Get BusAttachment"));
    Transport* transport = Transport::getInstance();
    return transport->getBusAttachment();
}

File: alljoyn/services/notification/java/NotificationService/src/org/alljoyn/ns/NotificationService.java
/******************************************************************************
 * Copyright AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 ******************************************************************************/

package org.alljoyn.ns;

import org.alljoyn.bus.BusAttachment;
import org.alljoyn.ns.commons.GenericLogger;
import org.alljoyn.ns.commons.NativePlatform;
import org.alljoyn.ns.commons.NativePlatformFactory;
import org.alljoyn.ns.commons.NativePlatformFactoryException;
import org.alljoyn.ns.transport.Transport;
import org.alljoyn.services.common.PropertyStore;


/**
 * The entry point of the Notification service <br>
 * The class is a singleton
 */
public class NotificationService  {
	private static final String TAG = "ioe" + NotificationService.class.getSimpleName();
	
	/**
	 * Initialize the class object
	 */
	private static final NotificationService SELF = new NotificationService();
	
	/**
	 * The protocol version that is sent in the {@link Notification} message
	 */
	public static final int PROTOCOL_VERSION = 2;
	
	/**
	 * Reference to transport object
	 */
	private Transport transport;
	
	/**
	 * Reference to native platform object
	 */
	private NativePlatform nativePlatform;
	
	/**
	 * Constructor
	 */
	private NotificationService() {}//constructor
	
	/**
	 * @return Returns the {@link NotificationService} object
	 */
	public static NotificationService getInstance() {
		return SELF;
	}//getInstance
	
	/**
	 * Establishes the notification sender service to later allow sending notifications 
	 * See {@link NotificationSender}
	 * @param bus The {@link BusAttachment} to be used by the service
	 * @param propertyStore Service configuration {@link PropertyStore}
	 * @return {@link NotificationSender}
	 * @throws NotificationServiceException Is thrown if an error occurred during the sender initialization 
	 */
	public NotificationSender initSend(BusAttachment bus, PropertyStore propertyStore) throws NotificationServiceException {
		init();
		
		GenericLogger logger = nativePlatform.getNativeLogger();

		if ( propertyStore == null ) {
			throw new NotificationServiceException("PropertyStore is NULL");
		}
		
		logger.debug(TAG, "Init Send invoked, calling Transport");
		transport.startSenderTransport(bus, propertyStore);
		
		logger.debug(TAG, "Creating and returning NotificationSender");
		return new NotificationSender();
	}//initSend

	/**
	 * Establishes the notification receiver service to later receive notifications. 
	 * See {@link NotificationReceiver}
	 * @param bus The {@link BusAttachment} to be used by the service
	 * @param receiver
	 * @throws NotificationServiceException if a malfunction occurred on building NotificationService
	 */
	public void initReceive(BusAttachment bus, NotificationReceiver receiver) throws NotificationServiceException {
		init();
		
		GenericLogger logger = nativePlatform.getNativeLogger();
		logger.debug(TAG, "Init Receive called, calling Transport");
		if ( receiver == null ) {
			throw new NotificationServiceException("NotificationReceiver interface should be implemented in order to receive notifications, received null pointer");
		}
		
		transport.startReceiverTransport(bus, receiver);
	}//initReceive

	/**
	 * Set a logger Object to enable logging capabilities
	 * @param logger GenericLogger object 
	 * @throws NotificationServiceException Is thrown if failed to set a logger
	 */	
	public void setLogger(GenericLogger logger) throws NotificationServiceException {
		init();
		
		if ( logger == null ) {
			throw new NullPointerException("Provided logger is NULL");
		}
		nativePlatform.setNativeLogger(logger);
	}//setLogger

	/**
	 * Shutdown notification service
	 * @throws NotificationServiceException
	 */
	public void shutdown() throws NotificationServiceException {
		init();
		transport.shutdown();
	}//shutdown

	/**
	 * Stops notification sender service
	 * @throws NotificationServiceException Throws if the service wasn't started previously 
	 */
	public void shutdownSender() throws NotificationServiceException {
		init();
		transport.shutdownSender();
	}//stopSender

	/**
	 * Stops notification receiver service
	 * @throws NotificationServiceException Throws if the service wasn't started previously 
	 */
	public void shutdownReceiver() throws NotificationServiceException {
		init();
		transport.shutdownReceiver();
	}//stopReceiver

	//=========================================================//
	
	/**
	 * Initializes the {@link NotificationService} object
	 * @throws NotificationServiceException Is thrown if failed to initialized the object
	 */
	private void init() throws NotificationServiceException {
		if ( nativePlatform != null &&  transport != null ) {
			return;
		}
		
		try {
			nativePlatform       = NativePlatformFactory.getPlatformObject();
			GenericLogger logger = nativePlatform.getNativeLogger();
			logger.info(TAG,  "Notification Service created");
			transport = Transport.getInstance();
		} catch (NativePlatformFactoryException npfe) {
			throw new NotificationServiceException("Failed to create Notification Service: " + npfe.getMessage());			
		} catch (Exception e) {
			throw new NotificationServiceException("Failed to create Notification Service: General Error");
		}
	}//init
	
}//NotificationServiceImpl
